moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1082 @@
+      pathIndices[stackSize - 1]++
+      peeked = PEEKED_NONE
+    }
+    val nextSource = JsonValueSource(source, prefix, state, valueSourceStackSize)
+    valueSource = nextSource
+    pushScope(JsonScope.STREAMING_VALUE)
+    return nextSource.buffer()
+  }
+
+  /**
+   * Returns the next character in the stream that is neither whitespace nor a part of a comment.
+   * When this returns, the returned character is always at `buffer.getByte(0)`.
+   */
+  private fun nextNonWhitespace(throwOnEof: Boolean): Int {
+    /*
+     * This code uses ugly local variable 'p' to represent the 'pos' field.
+     * Using locals rather than fields saves a few field reads for each
+     * whitespace character in a pretty-printed document, resulting in a
+     * 5% speedup. We need to flush 'p' to its field before any
+     * (potentially indirect) call to fillBuffer() and reread 'p' after
+     * any (potentially indirect) call to the same method.
+     */
+    var p = 0L
+    while (source.request(p + 1)) {
+      val c = buffer[p++].asChar()
+      when (c) {
+        '\n', ' ', '\r', '\t' -> continue
+      }
+      buffer.skip(p - 1)
+      when (c) {
+        '/' -> {
+          if (!source.request(2)) {
+            return c.code
+          }
+          checkLenient()
+          val peek = buffer[1]
+          return when (peek.asChar()) {
+            '*' -> {
+              // skip a /* c-style comment */
+              buffer.readByte() // '/'
+              buffer.readByte() // '*'
+              if (!skipToEndOfBlockComment()) {
+                throw syntaxError("Unterminated comment")
+              }
+              p = 0
+              continue
+            }
+            '/' -> {
+              // skip a // end-of-line comment
+              buffer.readByte() // '/'
+              buffer.readByte() // '/'
+              skipToEndOfLine()
+              p = 0
+              continue
+            }
+            else -> c.code
+          }
+        }
+        '#' -> {
+          // Skip a # hash end-of-line comment. The JSON RFC doesn't specify this behaviour, but it's
+          // required to parse existing documents.
+          checkLenient()
+          skipToEndOfLine()
+          p = 0
+        }
+        else -> return c.code
+      }
+    }
+    if (throwOnEof) {
+      throw EOFException("End of input")
+    }
+    return -1
+  }
+
+  private fun checkLenient() {
+    if (!lenient) {
+      throw syntaxError("Use JsonReader.setLenient(true) to accept malformed JSON")
+    }
+  }
+
+  /**
+   * Advances the position until after the next newline character. If the line is terminated by
+   * "\r\n", the '\n' must be consumed as whitespace by the caller.
+   */
+  private fun skipToEndOfLine() {
+    val index = source.indexOfElement(LINEFEED_OR_CARRIAGE_RETURN)
+    buffer.skip(if (index != -1L) index + 1 else buffer.size)
+  }
+
+  /** Skips through the next closing block comment. */
+  private fun skipToEndOfBlockComment(): Boolean {
```

Comment:
this is a drive by thing and it is not new to the port, but in case you want to solve it, this doesn't work if someone nested C-style comments. ex `[/* comment /* nested */ */ true]`

Reply:
Comments already aren't part of the standard so I'm not sure supporting that is necessary.